### PR TITLE
[GC] Add option G1BarrierSimple to use simple g1 post barrier

### PR DESCRIPTION
**Pull request by mmyxym**
Summary: Provide option G1BarrierSimple to use simple G1 post barrier for better mutator performance

Testing: CI pipeline

Reviewers: yude, yifeng

Issue: https://github.com/dragonwell-project/dragonwell11/issues/828

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 1]
```diff
@@ -482,7 +482,188 @@ class G1RefineCardClosure: public CardTableEntryClosure {
...
+    _scan_top = NULL;
+    return false;
+  }
+
+  size_t dirty_card_num() const {
+    return _dirty_card_num;
+  }
+};
+
+void G1RemSet::update_rem_set(G1ParScanThreadState* pss, uint worker_i, HeapRegionChunkClaimer* chunk_claimer) {
```
Comment (linade) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:666]: original patch has: assert(SafepointSynchronize::is_at_safepoint(), "Sanity");

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp [hunk 2]
```diff
@@ -2671,7 +2677,9 @@ class RegisterHumongousWithInCSetFastTestClosure : public HeapRegionClosure {
           if (g1h->is_in_closed_subset(ct->addr_for(card_ptr))) {
             if (*card_ptr != G1CardTable::dirty_card_val()) {
               *card_ptr = G1CardTable::dirty_card_val();
-              _dcq.enqueue(card_ptr);
+              if (!G1BarrierSimple) {
+                _dcq.enqueue(card_ptr);
+              }
```
Comment (linade) [inline src/hotspot/share/gc/g1/g1CollectedHeap.cpp:2682]: original patch has
```

if (G1BarrierSkipDCQ) {
--
  | 2649 | assert(G1ConcRefinementThreads == 0, "Concurrent refine should be disabled");
  | 2650 | assert(G1ConcRSLogCacheSize == 0, "Concurrent refine should be disabled");
  | 2651 | } else {
  | 2652 | _dcq.enqueue(card_ptr);
  | 2653 | }


```

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp [hunk 3]
```diff
@@ -120,6 +120,12 @@ void G1BarrierSet::invalidate(MemRegion mr) {
   volatile jbyte* byte = _card_table->byte_for(mr.start());
   jbyte* last_byte = _card_table->byte_for(mr.last());
   Thread* thr = Thread::current();
+  if (G1BarrierSimple) {
+    for (; byte <= last_byte; byte++) {
+      *byte = G1CardTable::dirty_card_val();
```
Comment (linade) [inline src/hotspot/share/gc/g1/g1BarrierSet.cpp:125]: a memset could be used: memset((void*)byte, G1CardTable::dirty_card_val(), last_byte - byte + 1);

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp [hunk 4]
```diff
@@ -658,6 +665,39 @@ bool G1BarrierSetC2::is_gc_barrier_node(Node* node) const {
 
 void G1BarrierSetC2::eliminate_gc_barrier(PhaseMacroExpand* macro, Node* node) const {
   assert(node->Opcode() == Op_CastP2X, "ConvP2XNode required");
+
```
Comment (linade) [inline src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:668]: does G1BarrierSetC2::step_over_gc_barrier need to change?

**Review by linade — COMMENTED** [review]
a few assertions are missing, e.g.,
```

void G1ConcurrentRefineThread::run_service() {
  assert(!G1CardMarkOnlyForTrackedRegion && !G1BarrierSkipDCQ, "Concurrent Refinement should be disabled");


```

File: src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp (hunk 4 above)
Comment (mmyxym) [reply src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp:668]: doesn't need to change

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp (hunk 3 above)
Comment (mmyxym) [reply src/hotspot/share/gc/g1/g1BarrierSet.cpp]: fixed

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp (hunk 2 above)
Comment (mmyxym) [reply src/hotspot/share/gc/g1/g1CollectedHeap.cpp:2682]: I think it's ok to not have the assertion.

File: src/hotspot/share/gc/g1/g1RemSet.cpp (hunk 1 above)
Comment (mmyxym) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:666]: I think it's ok.

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp [hunk 5]
```diff
@@ -120,6 +120,10 @@ void G1BarrierSet::invalidate(MemRegion mr) {
   volatile jbyte* byte = _card_table->byte_for(mr.start());
   jbyte* last_byte = _card_table->byte_for(mr.last());
   Thread* thr = Thread::current();
```
Comment (linade) [inline src/hotspot/share/gc/g1/g1BarrierSet.cpp:122]: is G1BarrierSet::write_ref_field_post_slow not used? perhaps add assert !G1BarrierSimple

Comment (linade) [reply src/hotspot/share/gc/g1/g1BarrierSet.cpp:122]: similarly for G1BarrierSetRuntime::write_ref_field_post_entry

Comment (mmyxym) [reply src/hotspot/share/gc/g1/g1BarrierSet.cpp:123]: Assertion added

**Review by leveretconey — COMMENTED** [review]
G1BarrierSimple is only supported on aarch64 and x86 now, so maybe it should be disabled on other platform

**Comment by linade** [comment]
PtrQueueSet::process_or_enqueue_complete_buffer also had changes in the original patch, wonder if it's missed

**Comment by mmyxym** [comment]
> PtrQueueSet::process_or_enqueue_complete_buffer also had changes in the original patch, wonder if it's missed

Add assertion to check dcqs empty

**Comment by mmyxym** [comment]
> G1BarrierSimple is only supported on aarch64 and x86 now, so maybe it should be disabled on other platform

Add the platform restriction.

**Review by linade — APPROVED** [review]

**Review by leveretconey — APPROVED** [review]

**Pull request merged**